crates/bevy_ecs/src/entity/index_map.rs

```diff
@@ -197,6 +351,471 @@ where
+    pub fn into_keys(self: Box<Self>) -> IntoKeys<V> {
+        IntoKeys(self.into_boxed_inner().into_keys(), PhantomData)
+    }
+
+    /// Return an iterator over mutable references to the the values of the map slice.
+    ///
+    /// Equivalent to [`map::Slice::values_mut`].
+    pub fn values_mut(&mut self) -> ValuesMut<'_, Entity, V> {
+        self.1.values_mut()
+    }
+
+    /// Return an owning iterator over the values of the map slice.
+    ///
+    /// Equivalent to [`map::Slice::into_values`].
+    pub fn into_values(self: Box<Self>) -> IntoValues<Entity, V> {
+        self.into_boxed_inner().into_values()
+    }
+}
+
+impl<V> Deref for Slice<V> {
+    type Target = map::Slice<Entity, V>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.1
+    }
+}
+
+impl<V: Debug> Debug for Slice<V> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        f.debug_tuple("Slice")
+            .field(&self.0)
+            .field(&&self.1)
+            .finish()
+    }
+}
+
+impl<V: Clone> Clone for Box<Slice<V>> {
+    fn clone(&self) -> Self {
+        // SAFETY: This a clone of a valid slice.
+        unsafe { Slice::from_boxed_slice_unchecked(self.as_boxed_inner().clone()) }
+    }
+}
+
+impl<V> Default for &Slice<V> {
+    fn default() -> Self {
+        // SAFETY: The source slice is empty.
+        unsafe { Slice::from_slice_unchecked(<&map::Slice<Entity, V>>::default()) }
+    }
+}
+
+impl<V> Default for &mut Slice<V> {
+    fn default() -> Self {
+        // SAFETY: The source slice is empty.
+        unsafe { Slice::from_slice_unchecked_mut(<&mut map::Slice<Entity, V>>::default()) }
+    }
+}
+
+impl<V> Default for Box<Slice<V>> {
+    fn default() -> Self {
+        // SAFETY: The source slice is empty.
+        unsafe { Slice::from_boxed_slice_unchecked(<Box<map::Slice<Entity, V>>>::default()) }
+    }
+}
+
+impl<V: Copy> From<&Slice<V>> for Box<Slice<V>> {
+    fn from(value: &Slice<V>) -> Self {
+        // SAFETY: This slice is a copy of a valid slice.
+        unsafe { Slice::from_boxed_slice_unchecked(value.1.into()) }
+    }
+}
+
+impl<V: Hash> Hash for Slice<V> {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.1.hash(state);
+    }
+}
+
+impl<'a, V> IntoIterator for &'a Slice<V> {
+    type Item = (&'a Entity, &'a V);
+    type IntoIter = Iter<'a, V>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        Iter(self.1.iter(), PhantomData)
+    }
+}
+
+impl<'a, V> IntoIterator for &'a mut Slice<V> {
+    type Item = (&'a Entity, &'a mut V);
+    type IntoIter = IterMut<'a, V>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        IterMut(self.1.iter_mut(), PhantomData)
+    }
+}
+
+impl<V> IntoIterator for Box<Slice<V>> {
+    type Item = (Entity, V);
+    type IntoIter = IntoIter<V>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        IntoIter(self.into_boxed_inner().into_iter(), PhantomData)
+    }
+}
+
+impl<V: PartialOrd> PartialOrd for Slice<V> {
```

Comment:
Can you `#[derive(PartialOrd, Ord, PartialEq, Eq, Hash)]` for these?  (And similarly for `index_set::Slice`)